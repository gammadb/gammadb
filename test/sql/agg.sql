--
-- Regression Tests for Custom Plan APIs
--

create extension gammadb;
-- construction of test data

CREATE TABLE t1 (
    c1   int,
    c2   int,
    c3   int,
    c4   int
) using gamma;
INSERT INTO t1 SELECT i%7,i%13,i%19,i%23 from generate_series(1,10000) i;
VACUUM ANALYZE t1;

set work_mem = '100MB';

EXPLAIN (COSTS OFF) SELECT count(c1) FROM t1;
SELECT count(c1) FROM t1;

set enable_hashagg to on;

EXPLAIN (COSTS OFF) SELECT c4, sum(c2) FROM t1 group by c4;
SELECT c4, sum(c2)  FROM t1 group by c4;

EXPLAIN (COSTS OFF) SELECT c4, sum(c2) FROM t1 where c3 >3 group by c4;
SELECT c4, sum(c2)  FROM t1 where c3 >3 group by c4;

EXPLAIN (COSTS OFF) SELECT sum(c1), count(distinct c4) FROM t1 group by c3,c2;
SELECT sum(c1), count(distinct c4) FROM t1 group by c3,c2;

EXPLAIN (COSTS OFF) SELECT sum(c1), count(distinct c4) FROM t1 group by rollup(c3,c2);
SELECT sum(c1), count(distinct c4) FROM t1 group by rollup(c3,c2);

set enable_hashagg to off;

EXPLAIN (COSTS OFF) SELECT c4, sum(c2) FROM t1 group by c4;
SELECT c4, sum(c2)  FROM t1 group by c4;

EXPLAIN (COSTS OFF) SELECT c4, sum(c2) FROM t1 where c3 >3 group by c4;
SELECT c4, sum(c2)  FROM t1 where c3 >3 group by c4;

reset enable_hashagg;

drop table t1;

drop extension gammadb;
