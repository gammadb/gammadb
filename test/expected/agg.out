--
-- Regression Tests for Custom Plan APIs
--
create extension gammadb;
-- construction of test data
CREATE TABLE t1 (
    c1   int,
    c2   int,
    c3   int,
    c4   int
) using gamma;
INSERT INTO t1 SELECT i%7,i%13,i%19,i%23 from generate_series(1,10000) i;
VACUUM ANALYZE t1;
set work_mem = '100MB';
EXPLAIN (COSTS OFF) SELECT count(c1) FROM t1;
                  QUERY PLAN                   
-----------------------------------------------
 Custom Scan (gamma_vec_agg)
   ->  Aggregate
         ->  Custom Scan (gamma_vec_tablescan)
               ->  Seq Scan on t1
(4 rows)

SELECT count(c1) FROM t1;
 count 
-------
 10000
(1 row)

set enable_hashagg to on;
EXPLAIN (COSTS OFF) SELECT c4, sum(c2) FROM t1 group by c4;
                  QUERY PLAN                   
-----------------------------------------------
 Custom Scan (gamma_vec_agg)
   ->  HashAggregate
         Group Key: c4
         ->  Custom Scan (gamma_vec_tablescan)
               ->  Seq Scan on t1
(5 rows)

SELECT c4, sum(c2)  FROM t1 group by c4;
 c4 | sum  
----+------
  0 | 2607
 17 | 2605
  7 | 2610
 11 | 2621
 13 | 2607
 12 | 2614
  6 | 2604
 20 | 2605
  9 | 2609
  1 | 2613
 15 | 2606
 19 | 2600
  8 | 2616
 10 | 2615
  3 | 2599
 14 | 2613
  5 | 2611
 16 | 2599
 18 | 2611
  2 | 2606
  4 | 2605
 21 | 2610
 22 | 2602
(23 rows)

EXPLAIN (COSTS OFF) SELECT c4, sum(c2) FROM t1 where c3 >3 group by c4;
                        QUERY PLAN                         
-----------------------------------------------------------
 Custom Scan (gamma_vec_agg)
   ->  HashAggregate
         Group Key: c4
         ->  Custom Scan (gamma_vec_tablescan)
               ->  Seq Scan on t1
                     Filter: (c3 OPERATOR(pg_catalog.>) 3)
(6 rows)

SELECT c4, sum(c2)  FROM t1 where c3 >3 group by c4;
 c4 | sum  
----+------
  0 | 2054
 17 | 2043
  7 | 2071
 11 | 2069
 13 | 2065
 12 | 2064
  6 | 2069
 20 | 2054
  9 | 2052
  1 | 2055
 15 | 2054
 19 | 2054
  8 | 2063
 10 | 2067
  3 | 2044
 14 | 2066
  5 | 2067
 16 | 2042
 18 | 2057
  2 | 2056
  4 | 2065
 21 | 2067
 22 | 2054
(23 rows)

EXPLAIN (COSTS OFF) SELECT sum(c1), count(distinct c4) FROM t1 group by c3,c2;
                              QUERY PLAN                               
-----------------------------------------------------------------------
 Custom Scan (gamma_vec_agg)
   ->  GroupAggregate
         Group Key: c3, c2
         ->  Custom Scan (gamma_vec_sort)
               ->  Sort
                     Sort Key: c3 USING <, c2 USING <
                     ->  Custom Scan (gamma_vec_result)
                           ->  Result
                                 ->  Custom Scan (gamma_vec_tablescan)
                                       ->  Seq Scan on t1
(10 rows)

SELECT sum(c1), count(distinct c4) FROM t1 group by c3,c2;
 sum | count 
-----+-------
 121 |    23
 120 |    23
 119 |    23
 118 |    23
 124 |    23
 120 |    23
 123 |    23
 124 |    23
 116 |    23
 122 |    23
 126 |    23
 122 |    23
 125 |    23
 124 |    23
 120 |    23
 118 |    23
 124 |    23
 116 |    23
 123 |    23
 126 |    23
 122 |    23
 122 |    23
 121 |    23
 120 |    23
 125 |    23
 121 |    23
 120 |    23
 123 |    23
 126 |    23
 116 |    23
 122 |    23
 121 |    23
 122 |    23
 125 |    23
 121 |    23
 120 |    23
 119 |    23
 118 |    23
 124 |    23
 123 |    23
 126 |    23
 122 |    23
 125 |    23
 121 |    23
 120 |    23
 119 |    23
 121 |    23
 124 |    23
 120 |    23
 118 |    23
 124 |    23
 116 |    23
 121 |    23
 122 |    23
 125 |    23
 121 |    23
 124 |    23
 119 |    23
 118 |    23
 124 |    23
 120 |    23
 123 |    23
 126 |    23
 116 |    23
 122 |    23
 120 |    23
 119 |    23
 121 |    23
 124 |    23
 120 |    23
 123 |    23
 124 |    23
 116 |    23
 122 |    23
 126 |    23
 122 |    23
 125 |    23
 121 |    23
 119 |    23
 118 |    23
 124 |    23
 120 |    23
 123 |    23
 126 |    23
 122 |    23
 122 |    23
 121 |    23
 120 |    23
 125 |    23
 121 |    23
 124 |    23
 123 |    23
 124 |    23
 116 |    23
 122 |    23
 121 |    23
 122 |    23
 125 |    23
 121 |    23
 120 |    23
 119 |    23
 118 |    23
 124 |    23
 120 |    23
 126 |    23
 122 |    23
 122 |    23
 121 |    23
 120 |    23
 119 |    23
 121 |    23
 124 |    23
 120 |    23
 118 |    23
 124 |    23
 116 |    23
 123 |    23
 122 |    23
 125 |    23
 121 |    23
 120 |    23
 119 |    23
 118 |    23
 124 |    23
 120 |    23
 123 |    23
 126 |    23
 116 |    23
 122 |    23
 121 |    23
 119 |    23
 121 |    23
 124 |    23
 120 |    23
 118 |    23
 124 |    23
 116 |    23
 122 |    23
 126 |    23
 122 |    23
 125 |    23
 121 |    23
 120 |    23
 118 |    23
 124 |    23
 120 |    23
 123 |    23
 126 |    23
 116 |    23
 122 |    23
 121 |    23
 120 |    23
 125 |    23
 121 |    23
 124 |    23
 119 |    23
 124 |    23
 116 |    23
 122 |    23
 126 |    23
 122 |    23
 125 |    23
 121 |    23
 120 |    23
 119 |    23
 118 |    23
 124 |    23
 120 |    23
 123 |    23
 122 |    23
 122 |    23
 121 |    23
 120 |    23
 125 |    23
 121 |    23
 124 |    23
 119 |    23
 118 |    23
 124 |    23
 116 |    23
 123 |    23
 126 |    23
 125 |    23
 121 |    23
 120 |    23
 119 |    23
 118 |    23
 124 |    23
 120 |    23
 123 |    23
 124 |    23
 116 |    23
 122 |    23
 121 |    23
 122 |    23
 121 |    23
 124 |    23
 120 |    23
 118 |    23
 124 |    23
 116 |    23
 123 |    23
 126 |    23
 122 |    23
 122 |    23
 121 |    23
 120 |    23
 119 |    23
 124 |    23
 120 |    23
 123 |    23
 126 |    23
 116 |    23
 122 |    23
 121 |    23
 122 |    23
 125 |    23
 121 |    23
 120 |    23
 119 |    23
 118 |    23
 116 |    23
 122 |    23
 126 |    23
 122 |    23
 125 |    23
 121 |    23
 120 |    23
 119 |    23
 121 |    23
 124 |    23
 120 |    23
 118 |    23
 124 |    23
 122 |    23
 121 |    23
 120 |    23
 125 |    23
 121 |    23
 124 |    23
 119 |    23
 118 |    23
 124 |    23
 120 |    23
 123 |    23
 126 |    23
 116 |    23
(247 rows)

EXPLAIN (COSTS OFF) SELECT sum(c1), count(distinct c4) FROM t1 group by rollup(c3,c2);
                              QUERY PLAN                               
-----------------------------------------------------------------------
 Custom Scan (gamma_vec_agg)
   ->  GroupAggregate
         Group Key: c3, c2
         Group Key: c3
         Group Key: ()
         ->  Custom Scan (gamma_vec_sort)
               ->  Sort
                     Sort Key: c3 USING <, c2 USING <
                     ->  Custom Scan (gamma_vec_result)
                           ->  Result
                                 ->  Custom Scan (gamma_vec_tablescan)
                                       ->  Seq Scan on t1
(12 rows)

SELECT sum(c1), count(distinct c4) FROM t1 group by rollup(c3,c2);
  sum  | count 
-------+-------
   121 |    23
   120 |    23
   119 |    23
   118 |    23
   124 |    23
   120 |    23
   123 |    23
   124 |    23
   116 |    23
   122 |    23
   126 |    23
   122 |    23
   125 |    23
  1580 |    23
   124 |    23
   120 |    23
   118 |    23
   124 |    23
   116 |    23
   123 |    23
   126 |    23
   122 |    23
   122 |    23
   121 |    23
   120 |    23
   125 |    23
   121 |    23
  1582 |    23
   120 |    23
   123 |    23
   126 |    23
   116 |    23
   122 |    23
   121 |    23
   122 |    23
   125 |    23
   121 |    23
   120 |    23
   119 |    23
   118 |    23
   124 |    23
  1577 |    23
   123 |    23
   126 |    23
   122 |    23
   125 |    23
   121 |    23
   120 |    23
   119 |    23
   121 |    23
   124 |    23
   120 |    23
   118 |    23
   124 |    23
   116 |    23
  1579 |    23
   121 |    23
   122 |    23
   125 |    23
   121 |    23
   124 |    23
   119 |    23
   118 |    23
   124 |    23
   120 |    23
   123 |    23
   126 |    23
   116 |    23
   122 |    23
  1581 |    23
   120 |    23
   119 |    23
   121 |    23
   124 |    23
   120 |    23
   123 |    23
   124 |    23
   116 |    23
   122 |    23
   126 |    23
   122 |    23
   125 |    23
   121 |    23
  1583 |    23
   119 |    23
   118 |    23
   124 |    23
   120 |    23
   123 |    23
   126 |    23
   122 |    23
   122 |    23
   121 |    23
   120 |    23
   125 |    23
   121 |    23
   124 |    23
  1585 |    23
   123 |    23
   124 |    23
   116 |    23
   122 |    23
   121 |    23
   122 |    23
   125 |    23
   121 |    23
   120 |    23
   119 |    23
   118 |    23
   124 |    23
   120 |    23
  1575 |    23
   126 |    23
   122 |    23
   122 |    23
   121 |    23
   120 |    23
   119 |    23
   121 |    23
   124 |    23
   120 |    23
   118 |    23
   124 |    23
   116 |    23
   123 |    23
  1576 |    23
   122 |    23
   125 |    23
   121 |    23
   120 |    23
   119 |    23
   118 |    23
   124 |    23
   120 |    23
   123 |    23
   126 |    23
   116 |    23
   122 |    23
   121 |    23
  1577 |    23
   119 |    23
   121 |    23
   124 |    23
   120 |    23
   118 |    23
   124 |    23
   116 |    23
   122 |    23
   126 |    23
   122 |    23
   125 |    23
   121 |    23
   120 |    23
  1578 |    23
   118 |    23
   124 |    23
   120 |    23
   123 |    23
   126 |    23
   116 |    23
   122 |    23
   121 |    23
   120 |    23
   125 |    23
   121 |    23
   124 |    23
   119 |    23
  1579 |    23
   124 |    23
   116 |    23
   122 |    23
   126 |    23
   122 |    23
   125 |    23
   121 |    23
   120 |    23
   119 |    23
   118 |    23
   124 |    23
   120 |    23
   123 |    23
  1580 |    23
   122 |    23
   122 |    23
   121 |    23
   120 |    23
   125 |    23
   121 |    23
   124 |    23
   119 |    23
   118 |    23
   124 |    23
   116 |    23
   123 |    23
   126 |    23
  1581 |    23
   125 |    23
   121 |    23
   120 |    23
   119 |    23
   118 |    23
   124 |    23
   120 |    23
   123 |    23
   124 |    23
   116 |    23
   122 |    23
   121 |    23
   122 |    23
  1575 |    23
   121 |    23
   124 |    23
   120 |    23
   118 |    23
   124 |    23
   116 |    23
   123 |    23
   126 |    23
   122 |    23
   122 |    23
   121 |    23
   120 |    23
   119 |    23
  1576 |    23
   124 |    23
   120 |    23
   123 |    23
   126 |    23
   116 |    23
   122 |    23
   121 |    23
   122 |    23
   125 |    23
   121 |    23
   120 |    23
   119 |    23
   118 |    23
  1577 |    23
   116 |    23
   122 |    23
   126 |    23
   122 |    23
   125 |    23
   121 |    23
   120 |    23
   119 |    23
   121 |    23
   124 |    23
   120 |    23
   118 |    23
   124 |    23
  1578 |    23
   122 |    23
   121 |    23
   120 |    23
   125 |    23
   121 |    23
   124 |    23
   119 |    23
   118 |    23
   124 |    23
   120 |    23
   123 |    23
   126 |    23
   116 |    23
  1579 |    23
 29998 |    23
(267 rows)

set enable_hashagg to off;
EXPLAIN (COSTS OFF) SELECT c4, sum(c2) FROM t1 group by c4;
                        QUERY PLAN                         
-----------------------------------------------------------
 Custom Scan (gamma_vec_agg)
   ->  GroupAggregate
         Group Key: c4
         ->  Custom Scan (gamma_vec_sort)
               ->  Sort
                     Sort Key: c4 USING <
                     ->  Custom Scan (gamma_vec_tablescan)
                           ->  Seq Scan on t1
(8 rows)

SELECT c4, sum(c2)  FROM t1 group by c4;
 c4 | sum  
----+------
  0 | 2607
  1 | 2613
  2 | 2606
  3 | 2599
  4 | 2605
  5 | 2611
  6 | 2604
  7 | 2610
  8 | 2616
  9 | 2609
 10 | 2615
 11 | 2621
 12 | 2614
 13 | 2607
 14 | 2613
 15 | 2606
 16 | 2599
 17 | 2605
 18 | 2611
 19 | 2600
 20 | 2605
 21 | 2610
 22 | 2602
(23 rows)

EXPLAIN (COSTS OFF) SELECT c4, sum(c2) FROM t1 where c3 >3 group by c4;
                              QUERY PLAN                               
-----------------------------------------------------------------------
 Custom Scan (gamma_vec_agg)
   ->  GroupAggregate
         Group Key: c4
         ->  Custom Scan (gamma_vec_sort)
               ->  Sort
                     Sort Key: c4 USING <
                     ->  Custom Scan (gamma_vec_tablescan)
                           ->  Seq Scan on t1
                                 Filter: (c3 OPERATOR(pg_catalog.>) 3)
(9 rows)

SELECT c4, sum(c2)  FROM t1 where c3 >3 group by c4;
 c4 | sum  
----+------
  0 | 2054
  1 | 2055
  2 | 2056
  3 | 2044
  4 | 2065
  5 | 2067
  6 | 2069
  7 | 2071
  8 | 2063
  9 | 2052
 10 | 2067
 11 | 2069
 12 | 2064
 13 | 2065
 14 | 2066
 15 | 2054
 16 | 2042
 17 | 2043
 18 | 2057
 19 | 2054
 20 | 2054
 21 | 2067
 22 | 2054
(23 rows)

reset enable_hashagg;
drop table t1;
drop extension gammadb;
